org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/IssueLocationsView.java

```diff
@@ -0,0 +1,402 @@
+        Optional<ExtraPosition> current = Optional.of(root);
+        while (current.isPresent()) {
+          ExtraPosition currentValue = current.get();
+          flow.add(currentValue);
+          current = positions.stream().filter(p -> p.getParent() == currentValue).findFirst();
+        }
+        result.add(flow);
+      }
+      return result;
+    }
+
+    public IMarker getRoot() {
+      return rootMarker;
+    }
+
+    public List<List<ExtraPosition>> getFlows() {
+      return flows;
+    }
+
+  }
+  private class LocationsProvider implements ITreeContentProvider {
+
+    @Override
+    public Object[] getElements(Object inputElement) {
+      IMarker sonarlintMarker = (IMarker) inputElement;
+      if (sonarlintMarker.getAttribute(MarkerUtils.SONAR_MARKER_HAS_EXTRA_LOCATION_KEY_ATTR, false)) {
+        if (currentDoc == null) {
+          return new Object[] {"Please open the file containing this issue in an editor to see the flows"};
+        }
+        return new Object[] {new RootNode(sonarlintMarker, positions(p -> p.getMarkerId() == sonarlintMarker.getId()))};
+      } else {
+        return new Object[] {"No aditional locations associated with this issue"};
+      }
+    }
+
+    @Override
+    public Object[] getChildren(Object parentElement) {
+      if (parentElement instanceof RootNode) {
+        List<List<ExtraPosition>> flows = ((RootNode) parentElement).getFlows();
+        if (flows.size() > 1) {
+          AtomicInteger counter = new AtomicInteger(0);
+          return flows.stream().map(f -> f.size() > 1 ? new FlowRootNode("Flow " + counter.incrementAndGet(), f) : f.get(0)).toArray();
+        } else if (flows.size() == 1) {
+          return flows.get(0).toArray();
+        } else {
+          return new Object[0];
+        }
+      } else if (parentElement instanceof FlowRootNode) {
+        return ((FlowRootNode) parentElement).getPositions().toArray();
+      } else {
+        return new Object[0];
+      }
+    }
+
+    private List<ExtraPosition> positions(Predicate<? super ExtraPosition> filter) {
+      try {
+        return Arrays.asList(currentDoc.getPositions(MarkerUtils.SONARLINT_EXTRA_POSITIONS_CATEGORY))
+          .stream()
+          .map(p -> (ExtraPosition) p)
+          .filter(filter)
+          .collect(Collectors.toList());
+      } catch (BadPositionCategoryException e) {
+        SonarLintLogger.get().error("Unable to read positions", e);
+        return Collections.emptyList();
+      }
+    }
+
+    @Override
+    public Object getParent(Object element) {
+      return null;
+    }
+
+    @Override
+    public boolean hasChildren(Object element) {
+      return getChildren(element).length > 0;
+    }
+
+    @Override
+    public void dispose() {
+      // Do nothing
+    }
+
+    @Override
+    public void inputChanged(Viewer viewer, Object oldInput, Object newInput) {
+      // Do nothing
+    }
+
+  }
+
+  private static class LocationsViewer extends TreeViewer {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') This class has 7 parents which is greater than 5 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMaximumInheritanceDepth)